### PR TITLE
Error for negative start index in string section_issue#4871

### PR DESCRIPTION
**Pull request by AnshMehta1**
This pull request aims to fix issue https://github.com/lfortran/lfortran/issues/4871. Basically , a compile-time check is done for start index less than 1 in case of character array splicing or substring section and correspondingly , a semantic error is thrown for start index less than 1. Changes in codebase are made accordingly to check for character array and then for start index.

### CODE REVIEW

**Comment by AnshMehta1** [comment]
Changes made from previous PR :- 

1) A test which just compiles and doesn't throw error was added in integration tests with referrences updated ->
```
program main
    implicit none
    character(len=8) :: s
    s = "lfortran"
    print*, "s:", s(+(1+1):4)
 end program
```

2) Removed offset thing as it was supposed to handle 0 start index but the code with 0 start index was ,nevertheless , not going into that section ->
```
else if ( ASR::is_a<ASR::IntegerUnaryMinus_t>(*args[0].m_left) ) {
    ASR::expr_t* value = ASRUtils::expr_value(args[0].m_left);
    if ( value ) {
        int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(value)->m_n;
        if ( a < 0 ) {
            throw SemanticError("The first index in string section is less than 1", loc);
        } else {
            a_value = ASRUtils::EXPR((ASR::make_IntegerConstant_t(al, loc,
                            a, int_type)));
        }
    }
}
```

3) Added test for 0 start index using ASR node ->
```
if(l != nullptr){

    if (ASR::is_a<ASR::IntegerBinOp_t>(*l)) {

        ASR::IntegerBinOp_t *num_expr = ASR::down_cast<ASR::IntegerBinOp_t>(l);

        if(num_expr != nullptr){
            ASR::expr_t *left_bound = num_expr->m_left;

            if(left_bound != nullptr && ASR::is_a<ASR::IntegerConstant_t>(*left_bound)){

                ASR::IntegerConstant_t *left_bound_value = ASR::down_cast<ASR::IntegerConstant_t>(left_bound);

                if(left_bound_value != nullptr && ((left_bound_value->m_n) == 0)){
                    throw SemanticError("The first index in string section is less than 1", loc);
                }
            }
        }
    } 
}
```

File: tests/tests.toml [hunk 1]
```diff
@@ -4215,4 +4215,12 @@ asr_implicit_interface_and_typing = true
...
 filename = "errors/func_arg_array.f90"
+asr = true
+
+[[test]]
+filename = "errors/string_negative_start_index.f90"
+asr = true
+
+[[test]]
+filename = "../integration_tests/string_negative_start_index2.f90"
 asr = true
```
Comment (Pranavchiku) [inline tests/tests.toml:4226]: As the test runs well with LFortran, we don't need a reference test. Just register it in `CMakeLists.txt` inside `integration_tests` directory.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -4061,6 +4061,19 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                            if ( value ) {
+                                int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(value)->m_n;
+                                if ( a < 0 ) {
+                                    throw SemanticError("The first index in string section is less than 1", loc);
+                                } else {
+                                    a_value = ASRUtils::EXPR((ASR::make_IntegerConstant_t(al, loc,
+                                                    a, int_type)));
+                                }
+                            }
+                        }
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:4075]:
```suggestion
                        else if ( ASR::is_a<ASR::IntegerUnaryMinus_t>(*args[0].m_left) ) {
                            ASR::expr_t* value = ASRUtils::expr_value(args[0].m_left);
                            if ( value ) {
                                int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(value)->m_n - offset;
                                a_value = ASRUtils::EXPR((ASR::make_IntegerConstant_t(al, loc,
                                                    a, int_type)));
                            }
                        }
                         if ( a_value != nullptr ) {
                            int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(a_value))->m_n;
                            if ( a < 0 ) throw SemanticError("The first index in string section is less than 1", loc);
                         } 
```

I think this should work.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -4085,6 +4098,27 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+
+                                    ASR::IntegerConstant_t *left_bound_value = ASR::down_cast<ASR::IntegerConstant_t>(left_bound);
+                                
+                                    if(left_bound_value != nullptr && ((left_bound_value->m_n) == 0)){
+                                        throw SemanticError("The first index in string section is less than 1", loc);
+                                    }
+                                }
+                            }
+                        } 
+                    }
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:4120]: You may remove this block and apply the comment given above. Try if it works then that should be fine.

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
left few comments above

**Comment by AnshMehta1** [comment]
> left few comments above

Ok , So made the changes and now things are working well.

File: integration_tests/CMakeLists.txt [hunk 4]
```diff
@@ -1793,3 +1793,4 @@ RUN(NAME exit_02 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME struct_type_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 
 RUN(NAME func_parameter_type_02 LABELS gfortran) # function passed in other argument of function
+RUN(NAME string_negative_start_index2 LABELS gfortran)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:1796]: Test with LFortran too:
```suggestion
RUN(NAME string_negative_start_index2 LABELS gfortran llvm)
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 5]
```diff
@@ -4061,6 +4061,20 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                            if ( value ) {
+                                int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(value)->m_n - offset;
+                                a_value = ASRUtils::EXPR((ASR::make_IntegerConstant_t(al, loc,
+                                                    a, int_type)));
+                            }
+                        }
+
+                        if ( a_value != nullptr ) {
+                            int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(a_value)->m_n;
+                            if ( a < 0 ) throw SemanticError("The first index in string section is less than 1", loc);
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:4075]: This test condition s for less than 0, not less than 1. So either the condition or the error message should change.

Comment (AnshMehta1) [reply src/lfortran/semantics/ast_common_visitor.h:4075]: Actually , value of offset is 1 which I subtracted and thus , a < 0 condition is start index less than 1 . So, both the things are matching.

Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:4075]:
```suggestion
                            if ( a < -1 ) throw SemanticError("The first index in string section is less than 0", loc);
```

This should take care of it.

Comment (AnshMehta1) [reply src/lfortran/semantics/ast_common_visitor.h:4075]: I think , for 0 start index is not valid in fortran and so error must be thrown for start index <= 0 or < 1 and so both the error message and condition are correct . I also checked the error message on my system and they are working well . So , I don't think that we have to change anything.

**Review by Pranavchiku — APPROVED** [review]

**Review by certik — APPROVED** [review]

**Pull request merged**